compiler-cli/src/dependencies.rs

```diff
@@ -492,38 +607,307 @@ fn resolve_versions<Telem: Telemetry>(
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    let error = Err(Error::DependencyResolutionFailed(
+        "wrong_name was expected but hello_world was found".to_string(),
+    ));
+    assert_eq!(result, error)
+}
+
+#[test]
+fn provide_existing_package() {
+    let mut provided = HashMap::new();
+
+    let result = provide_local_package(
+        "hello_world".into(),
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    assert_eq!(
+        result,
+        Ok(hexpm::version::Range::new("== 0.1.0".to_string()))
+    );
+
+    let result = provide_local_package(
+        "hello_world".into(),
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    assert_eq!(
+        result,
+        Ok(hexpm::version::Range::new("== 0.1.0".to_string()))
+    );
+}
+
+#[test]
+fn provide_conflicting_package() {
+    let mut provided = HashMap::new();
+
+    let result = provide_local_package(
+        "hello_world".into(),
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    assert_eq!(
+        result,
+        Ok(hexpm::version::Range::new("== 0.1.0".to_string()))
+    );
+
+    let result = provide_package(
+        "hello_world".into(),
+        &Path::new("../test/other"),
+        ProvidedPackageSource::Local {
+            path: Path::new("../test/other").to_path_buf(),
+        },
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    assert!(result.is_err()); // Error contains canonical path, so we cannot assert against the actual error value
+}
+
+#[test]
+fn provided_is_absolute() {
+    let mut provided = HashMap::new();
+    let result = provide_local_package(
+        "hello_world".into(),
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "subpackage".into()],
+    );
+    assert_eq!(
+        result,
+        Ok(hexpm::version::Range::new("== 0.1.0".to_string()))
+    );
+    let package = provided.get("hello_world").unwrap().clone();
+    if let ProvidedPackageSource::Local { path } = package.source {
+        assert!(path.is_absolute())
+    } else {
+        panic!("Provide_local_package provided a package that is not local!")
+    }
+}
+
+#[test]
+fn provided_recursive() {
+    let mut provided = HashMap::new();
+    let result = provide_local_package(
+        "hello_world".into(),
+        &Path::new("../test/hello_world"),
+        &mut provided,
+        &mut vec!["root".into(), "hello_world".into(), "subpackage".into()],
+    );
+    assert_eq!(
+        result,
+        Err(Error::ProvidedDependencyCycle(
+            "hello_world".to_string(),
+            "root -> hello_world -> subpackage".to_string()
```

Comment:
This should be a vector which then gets printed using the cycle printing code in the error module.

Reply:
Just used the existing error code for package cycles.